## Importing Library & Loading Dataset

In [42]:
import pandas as pd
import numpy as np
df= pd.read_csv("../raw_data/customer_shopping_behavior.csv")

Initial shape: (10000, 26)


## Data Inspection

In [80]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 33 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Transaction ID          10000 non-null  object        
 1   Customer ID             10000 non-null  object        
 2   Purchase Date           10000 non-null  datetime64[ns]
 3   Year                    10000 non-null  int32         
 4   Month                   10000 non-null  int32         
 5   Day                     10000 non-null  int32         
 6   Weekday                 10000 non-null  object        
 7   Age                     10000 non-null  int64         
 8   Age Group               10000 non-null  object        
 9   Gender                  10000 non-null  object        
 10  Cities                  10000 non-null  object        
 11  Region                  10000 non-null  object        
 12  Online/Offline          10000 non-null  object 

In [81]:
df.describe()

,Purchase Date,Year,Month,Day,Age,Quantity,Purchase Amount (₹),Discount (%),Shipping Charge (₹),Delivery Time (Days),Review Rating,Previous Purchases
count,10000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,2024-01-03 03:52:42.240000256,2023.496700,6.648800,15.446600,38.278100,1.425100,1889.882000,32.916300,41.49210,3.703600,3.874900,3.211400
min,2023-01-01 00:00:00,2023.000000,1.000000,1.000000,18.000000,1.000000,67.000000,0.000000,0.00000,0.000000,1.000000,0.000000
25%,2023-07-05 00:00:00,2023.000000,4.000000,8.000000,28.000000,1.000000,600.000000,12.000000,0.00000,1.000000,3.000000,0.000000
50%,2023-12-30 00:00:00,2023.000000,7.000000,15.000000,38.000000,1.000000,1207.000000,33.000000,0.00000,4.000000,4.000000,2.000000
75%,2024-07-05 00:00:00,2024.000000,10.000000,23.000000,49.000000,2.000000,2376.000000,52.000000,79.00000,6.000000,5.000000,4.000000
max,2024-12-31 00:00:00,2024.000000,12.000000,31.000000,59.000000,4.000000,22306.000000,70.000000,199.00000,8.000000,5.000000,28.000000
std,NaN,0.500014,3.527762,8.634941,12.133041,0.761346,2102.301836,21.440617,64.30923,2.726223,0.912871,3.951313


In [ ]:
print(f"Initial shape: {df.shape}")

In [44]:
print(df.head(30))

   Transaction ID Customer ID Purchase Date  Age  Gender       Location  \
0       TXN507808  CUST102534    2023-01-01   51    Male        Chennai   
1       TXN503830  CUST101256    2023-01-01   27  Female         Mumbai   
2       TXN501496  CUST100490    2023-01-01   57    Male      Hyderabad   
3       TXN501538  CUST100500    2023-01-01   22  Female           Pune   
4       TXN504265  CUST101399    2023-01-01   30    Male  Visakhapatnam   
5       TXN509250  CUST102952    2023-01-01   57    Male          Kochi   
6       TXN501814  CUST100596    2023-01-01   42    Male      Hyderabad   
7       TXN500031  CUST100007    2023-01-01   47    Male           Pune   
8       TXN501550  CUST100505    2023-01-01   37  Female      Ahmedabad   
9       TXN505122  CUST101684    2023-01-01   33    Male  Visakhapatnam   
10      TXN501568  CUST100510    2023-01-01   59    Male        Chennai   
11      TXN508080  CUST102612    2023-01-01   25    Male         Mumbai   
12      TXN501212  CUST10

### Duplicate Record Check

In [ ]:
df.duplicated().sum()

0

- Note:
    - Duplicates Not Found. 

## Data Cleaning

- 1. Null Value Check and Handling.
- 2. Renaming column.
- 3. Dtype Correction.

### 1. Null Value Check and Handling.

In [48]:
# Null Value Check:
df.isnull().sum().sort_values(ascending=False)

Festival/Sale             6788
Online Store              2249
Delivery Speed            2249
Size                      1035
Transaction ID               0
Purchase Amount (₹)          0
Previous Purchases           0
Return Status                0
Review Rating                0
Payment Method               0
Subscription Status          0
Delivery Time (Days)         0
Shipping Charge (₹)          0
Discount (%)                 0
Quantity                     0
Customer ID                  0
Color                        0
Brand                        0
Item Purchased               0
Category                     0
Online/Offline               0
Location                     0
Gender                       0
Age                          0
Purchase Date                0
Frequency of Purchases       0
dtype: int64

Note:
- Found nulls in Column:
    - A. Festival/Sale             6788
    - B. Online Store              2249
    - C. Delivery Speed            2249
    - D. Size                      1035

##### A. Null Value Handling for Column Festival/Sale

In [49]:
# 1. Column Unique Val check:
df['Festival/Sale'].unique()

array([nan, 'Republic Day Sale', 'Holi Sale', 'Amazon Great Summer Sale',
       'Independence Day Sale', 'Big Billion Days', 'Diwali Sale',
       'End of Season Sale'], dtype=object)

In [50]:
# 2. Null Val cross-relation check between 'Festival/Sale','Online Store':

df[df[['Festival/Sale','Online Store']].isnull().any(axis=1)][['Festival/Sale','Online Store']]

,Festival/Sale,Online Store
0,NaN,Flipkart
1,NaN,NaN
2,NaN,Flipkart
3,NaN,Myntra
4,NaN,Myntra
...,...,...
9987,End of Season Sale,NaN
9989,End of Season Sale,NaN
9991,End of Season Sale,NaN
9992,End of Season Sale,NaN


In [51]:
# 3. Null Value Handling:

df['Festival/Sale'] = df['Festival/Sale'].fillna('Regular Day')

In [52]:
# 4. Verifing Null Value Replacement:

print(df['Festival/Sale'].isna().sum())

0


##### B. Null Value Handling for Column Online Store:

In [ ]:
# 1. Null Value Handling:

df.loc[
    (df['Online/Offline'] == 'Offline') & 
    (df['Online Store'].isna()),
    'Online Store'
] = 'In-Store Purchase'

## Note: Where the Online/Offline was 'Offline' &  the Online Store was missing, I replaced the missing Online Store with "In-Store Purchase"

In [54]:
# 2. Verifing Null Value Replacement:

df[['Online/Offline', 'Online Store']].head(10)

,Online/Offline,Online Store
0,Online,Flipkart
1,Offline,In-Store Purchase
2,Online,Flipkart
3,Online,Myntra
4,Online,Myntra
5,Offline,In-Store Purchase
6,Online,Myntra
7,Online,Myntra
8,Offline,In-Store Purchase
9,Online,Amazon


In [85]:
# 3. Verifying any remaining Null:

df['Online Store'].isna().sum() 

0

##### C. Null Value Handling for Column Delivery Speed

In [55]:
# C(1). Null Value Handling:
df.loc[
    (df['Delivery Time (Days)'] == 0) & 
    (df['Delivery Speed'].isna()), 
    'Delivery Speed'
] = 'N/A (Offline)'

## Note: Where the delivery time was 0 days &  the Delivery Speed was missing, I replaced the missing Delivery Speed with "N/A (Offline)"

In [56]:
# C(2). Verifing Null Value Replacement:

print(df['Delivery Speed'].isna().sum())

0


##### D. Null Value Handling for Column Size

In [57]:
# 1. Column Unique Val check:
df.loc[df['Size'].isna(), 'Category'].unique()

array(['Accessories'], dtype=object)

Note:
- All Null values are from the same category.

In [58]:
# 2. Null Val cross-relation check between Category, Item Purchased & Size:

df[df[['Category', 'Item Purchased','Size']].isnull().any(axis=1)][['Category', 'Item Purchased', 'Size']]

,Category,Item Purchased,Size
5,Accessories,Watch,NaN
33,Accessories,Watch,NaN
34,Accessories,Sunglasses,NaN
44,Accessories,Wallet,NaN
47,Accessories,Wallet,NaN
...,...,...,...
9935,Accessories,Belt,NaN
9954,Accessories,Watch,NaN
9974,Accessories,Watch,NaN
9988,Accessories,Sunglasses,NaN


Note:
- No specific sub-category relation found for the nulls. Thus, filling up with Unknown.

In [59]:
# 3. Null value handling:

df['Size'] = df['Size'].fillna('Unknown')

In [60]:
# 4.  Verifying Null Value Replacement:

print(df['Size'].isna().sum())

0


##### B. Renaming the Location column with Cities column.

In [61]:
df.rename(columns={'Location': 'Cities'}, inplace=True)

##### C. Convert Purchase Date to relevant Dtype

In [62]:
df['Purchase Date'] = pd.to_datetime(df['Purchase Date'], errors = 'coerce')

print(df['Purchase Date'].dtype)

datetime64[ns]


# Feature Engineering:
- A. Creating a New Region Column.
- B. Creating column for footwear and clothing size. 
- C. Extracting Parts of date.
- D. Creating Age group 

##### A. Creating a New Region Column:

In [63]:
# 1. All unique city names:

df['Cities'].unique()

array(['Chennai', 'Mumbai', 'Hyderabad', 'Pune', 'Visakhapatnam', 'Kochi',
       'Ahmedabad', 'Chandigarh', 'Delhi', 'Bangalore', 'Indore',
       'Jaipur', 'Kolkata', 'Patna', 'Lucknow', 'Nagpur', 'Noida',
       'Surat', 'Bhubaneswar', 'Gurgaon'], dtype=object)

In [64]:
# 2. Mapping the cities

region_map = {
    'Delhi': 'North',
    'Noida': 'North',
    'Gurgaon': 'North',
    'Chandigarh': 'North',
    'Jaipur': 'North',
    'Lucknow': 'North',

    'Chennai': 'South',
    'Hyderabad': 'South',
    'Visakhapatnam': 'South',
    'Kochi': 'South',
    'Bangalore': 'South',

    'Mumbai': 'West',
    'Pune': 'West',
    'Ahmedabad': 'West',
    'Nagpur': 'West',
    'Surat': 'West',

    'Kolkata': 'East',
    'Patna': 'East',
    'Bhubaneswar': 'East',

    'Indore': 'Central'
}

df['Region'] = df['Cities'].map(region_map)

In [65]:
# 3. Popping out the region column:

region = df.pop('Region')

# Finding the position of the Cities column:

city_loc = df.columns.get_loc('Cities')

# Inserting Region column immediately after Cities column:

df.insert(city_loc + 1, 'Region', region)


##### B. Creating column for Footwear & Clothing Size. 

In [66]:
# 1. Category column unique vals:
df['Category'].unique()

array(['Clothing', 'Accessories', 'Footwear'], dtype=object)

In [67]:
# 2. Footwear Size:
Footwear_Size = np.where(df['Category'] == 'Footwear', df['Size'], 'Not Applicable')

# 3. Clothing Size:
Clothing_Size = np.where(df['Category'] != 'Footwear', df['Size'], 'Not Applicable')

In [68]:
# 4. Finding Size position:

size_loc = df.columns.get_loc('Size')

# 5. Inserting columns immediately after Size column:

df.insert(size_loc + 1, 'Clothing_Size', Clothing_Size)

df.insert(size_loc + 2, 'Footwear_Size', Footwear_Size)

In [69]:
# 7. Dropping Size column:
df.drop( columns = ['Size'], inplace = True)

In [70]:
# 8. Column after changes:
df[['Category', 'Footwear_Size', 'Clothing_Size']].head(10)

,Category,Footwear_Size,Clothing_Size
0,Clothing,Not Applicable,M
1,Clothing,Not Applicable,L
2,Clothing,Not Applicable,XXL
3,Clothing,Not Applicable,XS
4,Accessories,Not Applicable,Free Size
5,Accessories,Not Applicable,Unknown
6,Footwear,10,Not Applicable
7,Clothing,Not Applicable,S
8,Clothing,Not Applicable,M
9,Clothing,Not Applicable,XS


##### C. Extracting Parts of Date

In [71]:
# 1. Finding the position of the Purchase Date column:
purchase_date_loc = df.columns.get_loc('Purchase Date')


# 2. Extracting Year and inserting it immediately after the Purchase Date column:
df.insert(purchase_date_loc + 1, 'Year', df['Purchase Date'].dt.year)


# 3. Extracting Month and inserting it immediately after the Year column:
df.insert(purchase_date_loc + 2, 'Month', df['Purchase Date'].dt.month)


# 4. Extracting Day and inserting it immediately after the Month column:
df.insert(purchase_date_loc + 3, 'Day', df['Purchase Date'].dt.day)


# 5. Extracting Weekday Name and inserting it immediately after the Day column:
df.insert(purchase_date_loc + 4, 'Weekday', df['Purchase Date'].dt.day_name())


D. Creating Age group:

In [72]:
def age_group(age):
    if age < 18:
        return 'Under 18'
    elif age <= 25:
        return '18-25'
    elif age <= 35:
        return '26-35'
    elif age <= 45:
        return '36-45'
    elif age <= 60:
        return '46-60'
    else:
        return 'Above 60'
    
# Finding the position of the age column:
age_loc = df.columns.get_loc('Age')

# Inserting the Age Group column immediately after the age column:
df.insert(age_loc + 1, 'Age Group', df['Age'].apply(age_group))

# Final Data Validation

In [83]:
df.duplicated().sum()

0

In [74]:
df.dtypes

Transaction ID                    object
Customer ID                       object
Purchase Date             datetime64[ns]
Year                               int32
Month                              int32
Day                                int32
Weekday                           object
Age                                int64
Age Group                         object
Gender                            object
Cities                            object
Region                            object
Online/Offline                    object
Online Store                      object
Category                          object
Item Purchased                    object
Brand                             object
Color                             object
Clothing_Size                     object
Footwear_Size                     object
Quantity                           int64
Purchase Amount (₹)                int64
Discount (%)                       int64
Festival/Sale                     object
Shipping Charge 

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 33 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Transaction ID          10000 non-null  object        
 1   Customer ID             10000 non-null  object        
 2   Purchase Date           10000 non-null  datetime64[ns]
 3   Year                    10000 non-null  int32         
 4   Month                   10000 non-null  int32         
 5   Day                     10000 non-null  int32         
 6   Weekday                 10000 non-null  object        
 7   Age                     10000 non-null  int64         
 8   Age Group               10000 non-null  object        
 9   Gender                  10000 non-null  object        
 10  Cities                  10000 non-null  object        
 11  Region                  10000 non-null  object        
 12  Online/Offline          10000 non-null  object 

In [84]:
print("Remaining null values:")
print(df.isnull().sum().sort_values(ascending=False))

print(f"\nFinal shape: {df.shape}")

Remaining null values:
Transaction ID            0
Color                     0
Previous Purchases        0
Return Status             0
Review Rating             0
Payment Method            0
Subscription Status       0
Delivery Time (Days)      0
Delivery Speed            0
Shipping Charge (₹)       0
Festival/Sale             0
Discount (%)              0
Purchase Amount (₹)       0
Quantity                  0
Footwear_Size             0
Clothing_Size             0
Brand                     0
Customer ID               0
Item Purchased            0
Category                  0
Online Store              0
Online/Offline            0
Region                    0
Cities                    0
Gender                    0
Age Group                 0
Age                       0
Weekday                   0
Day                       0
Month                     0
Year                      0
Purchase Date             0
Frequency of Purchases    0
dtype: int64

Final shape: (10000, 33)


In [78]:
df.head()

,Transaction ID,Customer ID,Purchase Date,Year,Month,Day,Weekday,Age,Age Group,Gender,...,Festival/Sale,Shipping Charge (₹),Delivery Speed,Delivery Time (Days),Subscription Status,Payment Method,Review Rating,Return Status,Previous Purchases,Frequency of Purchases
0,TXN507808,CUST102534,2023-01-01,2023,1,1,Sunday,51,46-60,Male,...,Regular Day,0,Same Day,1,Yes,Cash on Delivery,4,Not Returned,0,Fortnightly
1,TXN503830,CUST101256,2023-01-01,2023,1,1,Sunday,27,26-35,Female,...,Regular Day,0,N/A (Offline),0,No,UPI,5,Not Returned,0,Fortnightly
2,TXN501496,CUST100490,2023-01-01,2023,1,1,Sunday,57,46-60,Male,...,Regular Day,0,Standard,7,Yes,UPI,3,Not Returned,0,Rarely
3,TXN501538,CUST100500,2023-01-01,2023,1,1,Sunday,22,18-25,Female,...,Regular Day,0,Standard,6,Yes,Net Banking,4,Not Returned,0,Fortnightly
4,TXN504265,CUST101399,2023-01-01,2023,1,1,Sunday,30,26-35,Male,...,Regular Day,99,Express,2,No,Debit Card,4,Not Returned,0,Weekly


# Export Clean Data

In [79]:
df.to_csv("../clean_data/clean_customer_shopping_behavior_india.csv", index=False)

## Conclusion

- Performed an initial dataset inspection of the dataset using `.info()` and `.describe()`.

- Conducted duplicate record analysis.

- Identified a datatype inconsistency in the `Purchase Date` column.

- Renamed `Location` column to `Cities` to improve dataframe readablity.

- Identified and handled missing values in the following columns:
  - `Festival/Sale`
  - `Online Store`
  - `Delivery Speed`
  - `Size`

- Corrected datatype inconsistencies by converting the `Purchase Date` column from `object` to `datetime`.

- Engineered new business-relevant features, including:
  - Created a `Region` column immedietly after `Cities` using a mapping funtion.
  - Created a `Footwear_Size` & `Clothing_Size` column immedietly after `Size` then dropped the `Size` column.
  - Extracted `Year`, `Month`, `Day` and `Weekday` from the `Purchase Date` column and inserted them immedietly after it.
  - Created a `Age Group` column immedietly after `Age` using a user-defined (`def`) funtion

- Verified the data Validation and confirmed successful handling of missing values after preprocessing.

- Exported the cleaned dataset for further analysis.